{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"results/samples/HG002-hs37d5-300x.bam\" : No such file or directory\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] could not open alignment file `results/samples/HG002-hs37d5-300x.bam`: No such file or directory",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 120\u001b[0m\n\u001b[1;32m    117\u001b[0m limits \u001b[38;5;241m=\u001b[39m (\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m6\u001b[39m\u001b[38;5;124m\"\u001b[39m, coords[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m1\u001b[39m] \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1000\u001b[39m, coords[\u001b[38;5;241m1\u001b[39m][\u001b[38;5;241m2\u001b[39m] \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1000\u001b[39m, \u001b[38;5;241m1.0\u001b[39m)\n\u001b[1;32m    119\u001b[0m test \u001b[38;5;241m=\u001b[39m SubsampleBAM(in_file\u001b[38;5;241m=\u001b[39min_file, out_file\u001b[38;5;241m=\u001b[39mout_file, coords\u001b[38;5;241m=\u001b[39mcoords, limits\u001b[38;5;241m=\u001b[39mlimits)\n\u001b[0;32m--> 120\u001b[0m \u001b[43mtest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilter_mates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mplot\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[1], line 54\u001b[0m, in \u001b[0;36mSubsampleBAM.run\u001b[0;34m(self, filter_mates, plot)\u001b[0m\n\u001b[1;32m     49\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     50\u001b[0m \u001b[38;5;124;03mRun the seeded subsampling procedure\u001b[39;00m\n\u001b[1;32m     51\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     52\u001b[0m np\u001b[38;5;241m.\u001b[39mrandom\u001b[38;5;241m.\u001b[39mseed(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mseed)\n\u001b[0;32m---> 54\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mpysam\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mAlignmentFile\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43min_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m in_bam, pysam\u001b[38;5;241m.\u001b[39mAlignmentFile(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mout_file, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwb\u001b[39m\u001b[38;5;124m\"\u001b[39m, template\u001b[38;5;241m=\u001b[39min_bam) \u001b[38;5;28;01mas\u001b[39;00m out_bam:\n\u001b[1;32m     56\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m seg \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msegments:\n\u001b[1;32m     57\u001b[0m         reads \u001b[38;5;241m=\u001b[39m [read \u001b[38;5;28;01mfor\u001b[39;00m read \u001b[38;5;129;01min\u001b[39;00m in_bam\u001b[38;5;241m.\u001b[39mfetch(seg[\u001b[38;5;241m0\u001b[39m], seg[\u001b[38;5;241m1\u001b[39m], seg[\u001b[38;5;241m2\u001b[39m])]\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pysam/libcalignmentfile.pyx:751\u001b[0m, in \u001b[0;36mpysam.libcalignmentfile.AlignmentFile.__cinit__\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/pysam/libcalignmentfile.pyx:950\u001b[0m, in \u001b[0;36mpysam.libcalignmentfile.AlignmentFile._open\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] could not open alignment file `results/samples/HG002-hs37d5-300x.bam`: No such file or directory"
     ]
    }
   ],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "\n",
    "class SubsampleBAM:\n",
    "\n",
    "     def __init__(self, in_file: str, out_file: str, coords: list[tuple[str, int, int, float]], seed: int=42, limits: tuple[str, int, int, float]=(\"6\", 0, 171_115_067, 1.0)):\n",
    "        \"\"\"\n",
    "        Initialize class using input and output BAM files, genomic coordinates, subsetting coefficient, seed\n",
    "        \"\"\"\n",
    "\n",
    "        # File IO\n",
    "        self.in_file = in_file\n",
    "        self.out_file = out_file\n",
    "\n",
    "        # Output filename check\n",
    "        self.out_sorted_file = f\"{''.join(self.out_file.split('.')[:-1])}-sorted.bam\"\n",
    "        assert self.out_sorted_file != self.in_file,  f\"Change {out_file=} to not overwrite {in_file=}\"\n",
    "\n",
    "        # Outer limits for region plotted\n",
    "        self.limits = limits\n",
    "        assert len(limits) == 4, f\"Number of limits should be exactly 4, not {len(self.limits)}\"\n",
    "\n",
    "        # Define what regions to plot\n",
    "        self.coords = coords\n",
    "        for coord in self.coords:\n",
    "            assert coord[0] == self.limits[0], f\"Supplied chromosome numbers do not match {self.limits[0]=} != {self.coord[0]}\"\n",
    "            assert coord[2] > coord[1], f\"Start {coord[2]} value should be less than end {coord[1]} value\"\n",
    "            assert coord[3] >= 0.0 and coord[3] <= 1.0, f\"{coord[3]=} outside the range 0-1 (inclusive)\"\n",
    "\n",
    "        # Define which region to plot using what coefficient\n",
    "        self.segments = []\n",
    "        points = [(limits[0], limits[1], limits[1], limits[3])] + coords + [(limits[0], limits[2], limits[2], limits[3])]\n",
    "        for i, point in enumerate(points):\n",
    "            try:\n",
    "                self.segments.append((points[i][0], points[i][1] + 1, points[i+1][1], points[i][3]))\n",
    "            except IndexError as ie:\n",
    "                self.segments.pop()\n",
    "                self.segments.append((points[-1][0], points[i-1][1] + 1, points[-1][2], points[-1][3]))\n",
    "\n",
    "        assert len(self.segments) == len(self.coords) + 1, f\"Number of segments ({len(self.segments)}) != number of coords + 1 ({len(self.coords)})\"\n",
    "        \n",
    "        # Set seed\n",
    "        self.seed = seed\n",
    "\n",
    "     def run(self, filter_mates: bool=True, plot: bool=True):\n",
    "        \"\"\"\n",
    "        Run the seeded subsampling procedure\n",
    "        \"\"\"\n",
    "        np.random.seed(self.seed)\n",
    "\n",
    "        with pysam.AlignmentFile(self.in_file, \"rb\") as in_bam, pysam.AlignmentFile(self.out_file, \"wb\", template=in_bam) as out_bam:\n",
    "\n",
    "            for seg in self.segments:\n",
    "                reads = [read for read in in_bam.fetch(seg[0], seg[1], seg[2])]\n",
    "                print(f\"Number of reads in region: {len(reads)}\")\n",
    "                \n",
    "                subset_reads = np.random.choice(reads, size=math.ceil(seg[3] * len(reads)), replace=False)\n",
    "\n",
    "                if filter_mates:\n",
    "                    subset_mates = []\n",
    "                    nomate_count = 0\n",
    "                    for read in subset_reads:\n",
    "                        try:\n",
    "                            subset_mates.append(in_bam.mate(reads))\n",
    "                        except:\n",
    "                            nomate_count += 1\n",
    "\n",
    "                    print(f\"No-mates read count: {nomate_count}\\n\")\n",
    "\n",
    "                    for read in subset_reads:\n",
    "                        if read not in subset_mates:\n",
    "                            out_bam.write(read)\n",
    "                \n",
    "                else:\n",
    "                    for read in subset_reads:\n",
    "                        out_bam.write(read)\n",
    "            \n",
    "\n",
    "        pysam.sort(self.out_file, \"-o\", self.out_sorted_file)\n",
    "        pysam.index(self.out_sorted_file)\n",
    "\n",
    "        if plot:\n",
    "            self.plot()\n",
    "\n",
    "     def plot(self):\n",
    "        \"\"\"\n",
    "        Plot in and out BAM file depths\n",
    "        \"\"\"\n",
    "        with pysam.AlignmentFile(self.in_file, \"rb\") as in_bam, \\\n",
    "            pysam.AlignmentFile(self.out_sorted_file, \"rb\") as out_bam:\n",
    "\n",
    "            in_pileup = [(p.pos, p.n) for p in in_bam.pileup(self.limits[0], self.limits[1], self.limits[2])]\n",
    "            out_pileup = [(p.pos, p.n) for p in out_bam.pileup(self.limits[0], self.limits[1], self.limits[2])]\n",
    "\n",
    "            fig, ax = plt.subplots(layout=\"constrained\")\n",
    "\n",
    "            ax.plot([p[0] for p in in_pileup], [p[1] for p in in_pileup], label=f\"{self.in_file.split('/')[-1]}\")\n",
    "            ax.plot([p[0] for p in out_pileup], [p[1] for p in out_pileup], label=f\"{self.out_sorted_file.split('/')[-1]}\")\n",
    "\n",
    "            ax.grid(visible=True, linestyle=\"--\", linewidth=1)\n",
    "            ax.ticklabel_format(useOffset=False, style='plain')\n",
    "            ax.set_title(f\"Coverage across chr{self.limits[0]}:{self.limits[1]}-{self.limits[2]}\")\n",
    "            ax.set_xlabel(\"Chromosomal coordinate\")\n",
    "            ax.set_ylabel(\"Depth of coverage\")\n",
    "            ax.legend()\n",
    "\n",
    "            plt.show()\n",
    "\n",
    "     \n",
    "# in_file = \"temp-downsample/AH8VC6ADXX-sorted.bam\"\n",
    "in_file = \"results/samples/HG002-hs37d5-300x.bam\"\n",
    "out_file = \"temp-downsample/out.bam\"\n",
    "coords = [(\"6\", 29_909_037, 29_913_661, 0.1), (\"6\", 29_913_662, 29_918_286, 0.1)]\n",
    "limits = (\"6\", coords[0][1] - 1000, coords[1][2] + 1000, 1.0)\n",
    "\n",
    "test = SubsampleBAM(in_file=in_file, out_file=out_file, coords=coords, limits=limits)\n",
    "test.run(filter_mates=True, plot=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
