# subsample-reads
Subsample reads is a Python tool that leverages the `pysam` package to subset fractions of reads across multiple defined chromosomal intervals in a single contig using a BED file that can also be generated by the tool.

Other tools that offer similar functionality such as `samtools view` or `GATK DownsampleSam` only process a single chromosomal region at a time, require multiple command line invocations and offer downsampling based on only a single fraction. What sets `subsample-reads` apart from other tools is the ability to specify a subsampling distribution across an entire contig and output a single subsetted BAM file. Other functions of the tool allow users to generate a sampling distribution from an existing, different BAM file, and to plot the results (or any BAM file).

During subsampling, `subsample-reads` randomly attributes reads that lie on the boundary of two intervals to either interval. No other checks are currently applied.

## Installing dependencies:
1. Create a Python virtual environment:
    
    `python -m venv venv`
1. Activate environment:
    
    `source venv/bin/activate`
1. Run `pip`:

    `pip install -r requirements.txt`
1. (Optional) Disable environment when done:

    `deactivate`

## Running:

### Prerequisites:

1. Install Python (tested on 3.11.4) dependencies and activate environment

#### `map`
2. One or more BAM files to map a sampling distribution from
3. A valid contig, start and end coordinates for the region to map
4. Interval size or count to set up when mapping read coverage across defined region

#### `sample`
2. One BAM file to sample according to a sampling distribution
3. BED file specifying sampling distribution (possibly created by `map`)
4. (Optional) Integer seed to use for sampling (for reproducibility)

#### `plot`
2. One or more BAM files to plot in the provided regions
3. BED file specifying sampling distribution (possibly created by `map`, only contig and start-end coordinates are used)

---

### Input file specifications:

1. BAM files with the following attributes:
    * Sorted, indexed
    * Contain reads mapped to specified regions
2. BED file with the following attributes (`map` will automatically produce a BED file in the following format):
    * No header row
    * Columns representing `contig`, `begin`, `end`, `fraction`, and `read_count` columns in a tab-separated format, not unlike conventional BED files. 
        - `contig` is any contig name featured in the input BAM file header ("chrN" and just "N" contig name formats are handled internally by `sample`)
        - `begin` is the starting chromosomal coordinate for a sampling interval in the `contig`
        - `end` is the ending coordinate for a sampling interval in the `contig`
        - `fraction` is a fraction in the interval [0.0, 1.0) for the region, all `fraction` values in BED file need to sum to approximately 1.0 (within a +- 0.05 margin).
        - `read_count` is an integer value specifying number of reads attributed to the region

---

### Execution:

Remember to source the Python virtual environment before running the tool:
```{python}
source venv/bin/activate
```

#### `map`
This command will produce a BED file with a sampling distribution from the BAM file based on read depths across the defined region.
```{python}
python -m subsample_reads map -i <input_BAM_files> -c <contig> -s <region_start> -e <region_end> (-w <window_size>/ -n <window_count>) -r <regions_BED_file>
```
- Providing multiple BAM files to map together will yield an averaged sampling distribution.

#### `sample`
This command will produce a BAM file with defined regions being sampled according to the distribution in the BED file.
```{python}
python -m subsample_reads sample -i <input_BAM_file> -r <regions_BED_file> -s <seed> -o <output_BAM_file>
```

#### `plot`
This command with one or more BAM files will produce a coverage plot of the regions in the BED file.
```{python}
python -m subsample_reads plot -i <input_BAM_files> -r <regions_BED_file> -o <output_plot_file>
```
- This function can technically plot any BAM file. A suggested way to use it to get a great plot is to specify:
    1. The unsampled BAM file
    2. BAM file(s) used to generate the sampling distribution in `map`
    3. The sampled BAM file
